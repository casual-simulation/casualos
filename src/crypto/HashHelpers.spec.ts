import { fromByteArray } from 'base64-js';
import { randomBytes } from 'tweetnacl';
import {
    createRandomPassword,
    hashPassword,
    verifyPassword,
    hashLowEntropyPasswordWithSalt,
    verifyPasswordAgainstHashes,
    hashHighEntropyPasswordWithSalt,
} from './HashHelpers';

describe('HashHelpers', () => {
    describe('hashPassword()', () => {
        it('should return a password hash', () => {
            const hash = hashPassword('password');
            expect(verifyPassword('password', hash)).toBe(true);
        });

        it('should throw if given a null password', () => {
            expect(() => {
                hashPassword(null);
            }).toThrow(
                new Error('Invalid password. Must not be null or undefined.')
            );
        });
    });

    describe('createRandomPassword()', () => {
        it('should return a password and hash', () => {
            const result = createRandomPassword();
            expect(verifyPassword(result.password, result.hash)).toBe(true);
        });
    });

    describe('hashLowEntropyPasswordWithSalt()', () => {
        it('should return a password hash', () => {
            const salt = fromByteArray(randomBytes(16));
            const hash = hashLowEntropyPasswordWithSalt('password', salt);
            expect(hash.startsWith('vH1.')).toBe(true);
            expect(verifyPasswordAgainstHashes('password', salt, [hash])).toBe(
                true
            );
        });

        it('should return a consistent hash', () => {
            const salt = 'GsVJeVGNV4+j1sjyfF13sg==';
            const hash = hashLowEntropyPasswordWithSalt('password', salt);
            expect(hash.startsWith('vH1.')).toBe(true);
            expect(verifyPasswordAgainstHashes('password', salt, [hash])).toBe(
                true
            );
            expect(hash).toMatchSnapshot();
        });

        it('should throw if given a null password', () => {
            expect(() => {
                hashLowEntropyPasswordWithSalt(null, '');
            }).toThrow(
                new Error('Invalid password. Must not be null or undefined.')
            );
        });

        it('should throw if given a null salt', () => {
            expect(() => {
                hashLowEntropyPasswordWithSalt('password', null);
            }).toThrow(
                new Error('Invalid salt. Must not be null or undefined.')
            );
        });
    });

    describe('hashHighEntropyPasswordWithSalt()', () => {
        it('should return a password hash', () => {
            const salt = fromByteArray(randomBytes(16));
            const hash = hashHighEntropyPasswordWithSalt('password', salt);
            expect(hash.startsWith('vH2.')).toBe(true);
            expect(verifyPasswordAgainstHashes('password', salt, [hash])).toBe(
                true
            );
        });

        it('should return a consistent hash', () => {
            const salt = 'GsVJeVGNV4+j1sjyfF13sg==';
            const hash = hashHighEntropyPasswordWithSalt('password', salt);
            expect(hash.startsWith('vH2.')).toBe(true);
            expect(verifyPasswordAgainstHashes('password', salt, [hash])).toBe(
                true
            );
            expect(hash).toMatchSnapshot();
        });

        it('should throw if given a null password', () => {
            expect(() => {
                hashHighEntropyPasswordWithSalt(null, '');
            }).toThrow(
                new Error('Invalid password. Must not be null or undefined.')
            );
        });

        it('should throw if given a null salt', () => {
            expect(() => {
                hashHighEntropyPasswordWithSalt('password', null);
            }).toThrow(
                new Error('Invalid salt. Must not be null or undefined.')
            );
        });
    });
});
