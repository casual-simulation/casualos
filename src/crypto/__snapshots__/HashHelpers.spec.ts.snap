// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashHelpers hashHighEntropyPasswordWithSalt() should return a consistent hash 1`] = `"vH2.rMXQq0dvdN9MYiSAudx1jbtNHlw4wrK5ATK1Bdg0alM="`;

exports[`HashHelpers hashLowEntropyPasswordWithSalt() should return a consistent hash 1`] = `"vH1.1JLjog5qJ0Uw5LB87Wnq2xjw2dcVMjQt6DYLVtyJC6Y="`;
